proposals/NNNN-hashable-enhancements.md

```diff
@@ -0,0 +1,925 @@
+
+There have been several attempts ([1][h1], [2][h2], [3][h3]) to fix `Hashable`
+by creating a new protocol that refines it:
+
+[h1]: https://github.com/apple/swift/blob/swift-4.1-branch/validation-test/stdlib/HashingPrototype.swift
+[h2]: https://github.com/attaswift/SipHash
+[h3]: https://gist.github.com/regexident/1b8e84974da2243e5199e760508d2d25
+
+```swift
+protocol Hashable {
+  var hashValue: Int { get }
+}
+
+protocol Hashable2: Hashable {
+  func hash(into hasher: inout Hasher)
+}
+
+extension Hashable2 {
+  var hashValue: Int {
+    var hasher = Hasher()
+    hash(into: &hasher)
+    return hasher.finalize()
+  }
+}
+```
+
+While this is a useful approach for external hashing packages, we
+believe it to be unsuitable for the standard library. Adding a new
+protocol would add a significant new source of user confusion about
+hashing, and it would needlessly expand the standard library's API
+surface area.
+
+The new protocol would need to have a new name, but `Hashable` already
+has the perfect name for a protocol representing hashable things -- so
+we'd need to choose an imperfect name for the "better" protocol.
+
+While deprecating a protocol requirement is a significant change, we
+believe it to be less harmful overall than leaving `Hashable`
+unchanged, or trying to have two parallel protocols for the same thing.
+
+Additionally, adding a second protocol would lead to complications
+with `Hashable` synthesis. It's also unclear if `Set` and `Dictionary`
+would be able to consistently use `Hasher` for their primary hashing
+API. (These problems are not unsolvable, but they may involve adding
+special one-off compiler support for the new protocol. For example, we
+may want to automatically derive `Hashable2` conformance for all types
+that implement `Hashable`.)
+
+### <a name="generic-hasher">Making `hash(into:)` generic over a `Hasher` protocol</a>
+
+It would be nice to allow Swift programmers to define their own hash
+functions, and to plug them into any `Hashable` type:
+
+```swift
+protocol Hasher {
+  func append(bits: Int)
+  func append(bits: UInt)
+  /// etc.
+}
+protocol Hashable {
+  func hash<H: Hasher>(into hasher: inout H)
+}
+```
+
+However, we believe this would add an unnnecessary degree of freedom.
+In particular, we do not foresee a need for adding support for custom
+hashers in `Set` and `Dictionary`. On the other hand, there are
+distinct advantages to standardizing on a single, high-quality hash
+function:
+
+* Adding a generic type parameter to `hash(into:)` would complicate
+    the `Hashable` API.
+* By supporting only a single `Hasher`, we can concentrate our efforts
+    on making sure it runs fast. For example, we know that the
+    standard hasher's opaque mutating functions won't ever perform any
+    retain/release operations, or otherwise mutate any of the
+    reference types we may encounter during hashing; describing this
+    fact to the compiler may enable optimizations that would not
```

Comment:
I'm all for qualifying uncertain statements with "may", but in this case we have can make a statement with certainty. How a function may interact with the rest of the system, especially in relation to the lifetime of non-parameter objects, affects how conservative (i.e. slow) the Swift compiler has to be to ensure memory safety.